lib/OpenQA/Utils.pm

```diff
@@ -224,71 +224,68 @@ sub needle_info {
 }
 
 # logging helpers
-sub log_debug {
-    my ($msg, $channel) = @_;
 
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->debug($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->debug($msg);
-    }
-    else {
-        STDOUT->printflush("[DEBUG] $msg\n");
-    }
+sub log_debug {
+    _log_msg('debug', @_);
 }
 
 sub log_info {
-    my ($msg, $channel) = @_;
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->info($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->info($msg);
-    }
-    else {
-        STDOUT->printflush("[INFO] $msg\n");
-    }
+    _log_msg('info', @_);
 }
 
 sub log_warning {
-    my ($msg, $channel) = @_;
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->warn($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->warn($msg);
-    }
-    else {
-        STDERR->printflush("[WARN] $msg\n");
-    }
+    _log_msg('warn', @_);
 }
 
 sub log_error {
-    my ($msg, $channel) = @_;
-    if ($channel && $channels{$channel}) {
-        $channels{$channel}->error($msg);
-    }
-    elsif ($app && $app->log) {
-        $app->log->error($msg);
-    }
-    else {
-        STDERR->printflush("[ERROR] $msg\n");
-    }
+    _log_msg('error', @_);
 }
 
 sub log_fatal {
-    my ($msg, $channel) = @_;
+    _log_msg('fatal', @_);
+    die $_[0];
+}
+
+sub _log_msg {
+    my ($level, $msg, %options) = @_;
+    my $code = <<"END";
+    my (\$msg, \%options) = \@_;
+    my \$log_default = \$options{default};
+
+    if(\$options{channels}) {
```

Comment:
I'd check if channels actually exist, before trying to log, and log fatal if they don't.

Reply:
@foursixnine The check is already being done in:

https://github.com/os-autoinst/openQA/pull/1502/files#diff-a8857daf563deea1776f5c009cd73a32R275